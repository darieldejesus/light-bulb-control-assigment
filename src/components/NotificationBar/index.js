import React from 'react';
import { Alert } from 'rendition';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { dismissError } from '../../stateManager/ducks/devices';

const mapStateToProps = ({ devices }) => ({
  error: devices.error,
});

const mapDispatchToProps = dispatch => ({
  onDismiss: () => dispatch(dismissError()),
});

/**
 * Component to show a box with the error message.
 *
 * @param {Object} props - Component props.
 * @param {string} props.error - Error message.
 * @param {Function} props.onDismiss - Callback to dismiss the error.
 * @returns {Object} As ReactElement.
 */
function NotificationBar({ error, onDismiss }) {
  return (
    error && (
      <Alert
        mb={2}
        py={0}
        danger
        onDismiss={onDismiss}
      >
        { error }
      </Alert>
    )
  );
}

NotificationBar.defaultProps = {
  error: false,
};

NotificationBar.propTypes = {
  error: PropTypes.oneOfType([
    PropTypes.string.isRequired,
    PropTypes.bool.isRequired,
  ]),
  onDismiss: PropTypes.func.isRequired,
};

export default connect(mapStateToProps, mapDispatchToProps)(NotificationBar);
