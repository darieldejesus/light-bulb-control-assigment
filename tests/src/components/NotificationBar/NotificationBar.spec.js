import React from 'react';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Provider as ReduxProvider } from 'react-redux';
import { Provider as RenditionProvider } from 'rendition';

import api from '../../../../src/helpers/ApiClient';
import clientMiddleware from '../../../../src/stateManager/middleware/client';
import NotificationBar from '../../../../src/components/NotificationBar';
import theme from '../../../../src/theme';

const middleware = [
  clientMiddleware(api),
  thunk,
];
const mockStore = configureMockStore(middleware);

describe('NotificationBar', () => {
  test('Should render the NotificationBar without error message', () => {
    const store = mockStore({
      devices: {
        error: false,
      },
    });
    const wrapper = shallow(
      <ReduxProvider store={store}>
        <NotificationBar />
      </ReduxProvider>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('Should render the NotificationBar with error message', () => {
    const store = mockStore({
      devices: {
        error: 'Ups! Something went wrong',
      },
    });
    const wrapper = shallow(
      <ReduxProvider store={store}>
        <NotificationBar />
      </ReduxProvider>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('Should call DISMISS message after clicking the close (x) button', () => {
    const store = mockStore({
      devices: {
        error: 'Ups! Something went wrong',
      },
    });
    const wrapper = mount(
      <ReduxProvider store={store}>
        <RenditionProvider theme={theme}>
          <NotificationBar />
        </RenditionProvider>
      </ReduxProvider>,
    );
    const expectedActions = [
      {
        type: 'resinio/devices/DISMISS',
      },
    ];
    wrapper.find('NotificationBar').props().onDismiss();
    expect(store.getActions()).toEqual(expectedActions);
  });
});
