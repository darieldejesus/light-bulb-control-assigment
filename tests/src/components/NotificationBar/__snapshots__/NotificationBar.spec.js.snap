// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationBar Should render the NotificationBar with error message 1`] = `<Connect(NotificationBar) />`;

exports[`NotificationBar Should render the NotificationBar without error message 1`] = `<Connect(NotificationBar) />`;
